server/client.go

```diff
@@ -1931,22 +1932,53 @@ func (c *client) addShadowSub(sub *subscription, im *streamImport, useFrom bool)
 
 // canSubscribe determines if the client is authorized to subscribe to the
 // given subject. Assumes caller is holding lock.
-func (c *client) canSubscribe(subject string) bool {
+func (c *client) canSubscribe(subject, queue string) bool {
+	//c.perms = new(permissions)
+	//c.perms.sub.deny = NewSublist(false)
+	//c.perms.sub.deny.Insert(&subscription{
+	//	subject: []byte("bar"),
+	//	queue:   []byte("foogroup"),
+	//})
+	//c.perms.sub.allow = NewSublist(false)
+	//c.perms.sub.allow.Insert(&subscription{
+	//	subject: []byte("foo"),
+	//	queue:   []byte("bargroup"),
+	//})
+
 	if c.perms == nil {
 		return true
 	}
 
 	allowed := true
 
-	// Check allow list. If no allow list that means all are allowed. Deny can overrule.
+	// Check allow list. If no allow list that means all are allowed. Deny can
+	// overrule.
 	if c.perms.sub.allow != nil {
 		r := c.perms.sub.allow.Match(subject)
-		allowed = len(r.psubs) != 0
+
+		if queue == "" && len(r.psubs) > 0 {
+			allowed = true
+		}
+
+		if queue != "" && len(r.qsubs) > 0 {
+			allowed = false
+			for i := 0; i < len(r.qsubs); i++ {
+				for j := 0; j < len(r.qsubs[i]); j++ {
+					q := string(r.qsubs[i][j].queue)
```

Comment:
qsubs[i] will all be same queue name. So just need to check first one, no need for iterator. But would check len() >= 1.

Reply:
Cool, got rid of the loop!